### ISSUE TITLE
Build for ARM64 version

### ISSUE THREAD

**Issue by sj6219** [basil00/Divert#236]
Please, create an arm64 version that works with ms surface pro x

**Comment by basil00** [comment]
This is a dup of #115, but the build system has been updated since then.

I'm all for an ARM version but have no way of testing it at the moment.

**Comment by sj6219** [comment]
After the build, I tested it to some extent.

But I understand it's hard to approve because you can't check it yourself.

I can't sign your driver program right now, so it's uncomfortable to use.
I'll try it when you release the signed arm version later.

Thank you.

**Comment by leiless** [comment]
Any update by now? Microsoft just introduced x64 emulation in preview for Windows 10 on ARM.
Sooner or later, Windows 10 ARM64 will be populated.
see:
https://blogs.windows.com/windows-insider/2020/12/10/introducing-x64-emulation-in-preview-for-windows-10-on-arm-pcs-to-the-windows-insider-program/

> I'm all for an ARM version but have no way of testing it at the moment.

@basil00, Just buy one. :-)

**Comment by sj6219** [comment]
It has been under development since yesterday.
Because x64 simulation method is different from x32 method, it seems to take time.

**Comment by sj6219** [comment]
See #237.
I would appreciate it if you digitally sign the attached arm64 version file.
[windivert-arm64.zip](https://github.com/basil00/Divert/files/5712978/windivert-arm64.zip)

**Comment by leiless** [comment]
@sj6219 Many thanks for what you've done, I'll rebuild it on my ARM64 machine.
@basil00, please consider support WinDivert on ARM64 Windows, and eveutally release a code signed package.

**Comment by sj6219** [comment]
To build and test on the arm machine
Please refer to the following documents

https://docs.microsoft.com/en-us/windows-hardware/drivers/download-the-wdk
https://www.howtogeek.com/167723/how-to-disable-driver-signature-verification-on-64-bit-windows-8.1-so-that-you-can-install-unsigned-drivers/
https://support.microsoft.com/en-us/windows/finding-your-bitlocker-recovery-key-in-windows-10-6b71ad27-0b89-ea08-f143-056f5ab347d6

**Comment by elmeyer** [comment]
I know +1s are generally discouraged, but I would also appreciate a signed build of the driver for ARM64.

**Comment by sanandchoubey** [comment]
@sj6219 Thanks for the wonderful work! Is the signed ARM64 build available like the non-ARM64 platform? If it is not available, should I download the code, build, and MS sign by myself?

**Comment by basil00** [comment]
> should I download the code, build, and MS sign by myself?

Yes, but please use this branch: https://github.com/basil00/Divert/tree/release-v2.2.1
This is the latest (v2.2.2) release.

**Comment by sj6219** [comment]
There are currently no signed driver files for arm.
So it's uncomfortable to use.

Source code here: https://github.com/sj6219/Divert/tree/ARM

built file here: https://github.com/sj6219/Divert/releases/download/ARM64/ARM64.zip

If you want to use, do the following:
Bcdedit.exe -set TESTSIGNING ON

**Comment by sanandchoubey** [comment]
> There are currently no signed driver files for arm. So it's uncomfortable to use.
> 
> Source code here: https://github.com/sj6219/Divert/tree/ARM
> 
> built file here: https://github.com/sj6219/Divert/releases/download/ARM64/ARM64.zip
> 
> If you want to use, do the following: Bcdedit.exe -set TESTSIGNING ON

Thanks @sj6219

**Comment by da-kami** [comment]
@sj6219 thanks for providing the original ARM build!
@sanandchoubey did you succeed in testing it?

I tried to test the build on ARM `Windows 11 Home` in test mode, but I keep getting this error when trying to run any of the examples included I get this error:

```
error: failed to open WinDivert handle (err = 3)
```

Here's what I did:

1. Download and unzip the ARM build from https://github.com/sj6219/Divert/releases/download/ARM64/ARM64.zip
2. Activate Test Mode on Windows
    1. Switch off secure boot
    2. Run `bcdedit /set testsigning on`, reboot (I see the Test Mode watermark now)
3. Install the certificate that the`sys` file was signed with
    1. Install in `Trusted Root Certification Authorities` store
    2. Install in `Trusted Publishers` store
4. Run `streamdump.exe 80` from the `ARM64` WinDivert folder (my understanding is that this example has been adopted)
    1.`error: failed to open the WinDivert device (3)`
    2. I tried running other examples / test, but get error: `error: failed to open WinDivert handle (err = 3)`

Am I missing something?
Has anyone succeeded with testing this build on Windows 11?

**Comment by da-kami** [comment]
> I tried to test the build on ARM `Windows 11 Home` in test mode, but I keep getting this error when trying to run any of the examples included I get this error:
> 
> ```
> error: failed to open WinDivert handle (err = 3)
> ```

Update: 
I managed to run the ARM build in test signing mode. 
The problem was that some old/incompatible version of WinDivert was still installed on the device. 
I ran `sc delete WinDivert` which gave me a `[SC] DeleteService SUCCESS` and after that it worked.
I was able to run `test.exe` as well and all tests passed.

**Comment by akahan** [comment]
I'd like to second this request and add some context that has changed since 2020.

   Windows on ARM is no longer a niche. With the Copilot+ PC generation (Snapdragon X Elite/Plus), ARM64 laptops have become one of the fastest-growing Windows segments, and we see a steadily increasing share of
   our users on these machines. WinDivert is the de-facto standard for user-space packet capture and modification on Windows, so many projects — VPN clients, firewalls, traffic filters, network analyzers, ours
   included — are currently forced to ship degraded functionality or no support at all on ARM64, simply because no ARM64 build of WinDivert exists.

   The real blocker is not compilation but driver signing. Building WinDivert.sys for ARM64 from source with the current WDK is straightforward, and community forks have done it. But a self-built driver is not
   loadable in practice:

   • Windows on ARM64 requires properly signed kernel drivers; test-signing demands enabling testsigning in BCD, which weakens security guarantees users shouldn't be asked to give up — and on many Snapdragon
     devices with Secure Boot it is a non-starter entirely.
   • Only the official signing chain already used for the x86/x64 releases can produce a driver that simply loads on stock systems.

   So we would like to ask: could you publish an official ARM64 build — a signed ARM64 WinDivert.sys together with the matching ARM64 WinDivert.dll — in the release assets, alongside the existing x86/x64 binaries?
   Even a one-off signed build of the current 2.2.x branch would unblock the whole ecosystem of downstream projects on ARM64.

   We're happy to help with whatever makes this easier: testing on real Snapdragon X hardware, verifying driver load and the capture/inject paths, or contributing CI scripts for the ARM64 target.

   Thank you for maintaining WinDivert — it remains the best tool of its kind on Windows.
